### PR TITLE
implement for loops

### PR DESCRIPTION
**Pull request by aalok-t**
### Description

Add support to parse for loops of the form `for (iter in lower_bound..upper_bound) { loop_body }` by transforming it into a corresponding while loop.

### Test

Refer to tests in `move-compiler/transactional-tests/tests/control_flow/`

### CODE REVIEW

File: third_party/move/move-compiler/src/parser/syntax.rs
Comment (aalok-t) [inline third_party/move/move-compiler/src/parser/syntax.rs:1207]: Currently we support let `for (iter = init; { spec { ... }; condition }; update) loop_body` instead.

**Comment by wrwg** [comment]
Just checking back, I think this is the version before we discussed in person?

**Comment by aalok-t** [comment]
Yes, I will push the updates soon.

**Comment by vineethk** [comment]
@aalok-t Can you let us know when we should review this PR? Maybe put a [not ready for review yet] or take down the PR in the meantime? Another alternative is to remove all reviewers and add them back again when this is ready for review.

**Comment by vineethk** [comment]
I ran the following transactional test, and it returned `12`: not `10` as one would expect.

```
//# publish
module 0x42::m {
    fun main(): u64 {
        let y = 0;
        for (i in 0..10) {
            y = y + 1;
        };
        y
    }
}

//# run 0x42::m::main
```

I think we need to add more tests: 
- ones with nested for loops
- nested for + while,
- `for (_ in 0..10) {}` leads to an unexpected (to the user) error, 
- modifying the iterator variable has different semantics than rust, we should have test cases to show this behavior, and defend why it is okay to diverge from the rust behavior.

**Comment by brmataptos** [comment]
Test I in loop

File: third_party/move/move-compiler/src/parser/syntax.rs
Comment (aalok-t) [reply third_party/move/move-compiler/src/parser/syntax.rs:1281]: To implement for loops at a parser level, we would need to include a flag (or some ghost variable) which cannot be referenced in the loop. One resolution is to create an error if the name of the ghost variable is referenced inside the loop.

Comment (aalok-t) [reply third_party/move/move-compiler/src/parser/syntax.rs:1281]: This is an inherent problem with having a flag. There are two ways to implement for loops without a flag. One is to convert it into a `loop`, which is a bit more involved, and the other is to add an if-statement for the loop body as done in the updated PR.

**Comment by aalok-t** [comment]
> Test I in loop

The tests `for_loop_increment_iter` and `for_loop_non_terminating` update the iterator in the loop body.

**Comment by aalok-t** [comment]
> I ran the following transactional test, and it returned `12`: not `10` as one would expect.
> 
> ```
> //# publish
> module 0x42::m {
>     fun main(): u64 {
>         let y = 0;
>         for (i in 0..10) {
>             y = y + 1;
>         };
>         y
>     }
> }
> 
> //# run 0x42::m::main
> ```
> 
> I think we need to add more tests:
> 
> * ones with nested for loops
> * nested for + while,
> * `for (_ in 0..10) {}` leads to an unexpected (to the user) error,
> * modifying the iterator variable has different semantics than rust, we should have test cases to show this behavior, and defend why it is okay to diverge from the rust behavior.

1. Fixed the error for the given program.
2. Added tests with nesting. 
3. The error for `for (_ in 0..10) {}` is `Invalid variable usage. Unbound variable '_'` as iterator is variable that is updated in the code. This is where we diverge from Rust. Do you suggest we add a clarifying error message for such a usage?
4. Unlike Rust, we do not consume the iterators in the loop and instead treat it like a mutable variable (as demonstrated in the updated tests).

File: third_party/move/move-compiler/src/parser/syntax.rs
Comment (brmataptos) [reply third_party/move/move-compiler/src/parser/syntax.rs:957]: I guess I wrote this code, but maybe it could use more docs.

File: third_party/move/move-compiler-v2/tests/checking/typing/while_condition.move
Comment (aalok-t) [reply third_party/move/move-compiler-v2/tests/checking/typing/while_condition.move:4]: No, I just noticed the that it was non-uniform. It does not affect the test.

**Comment by vineethk** [comment]
Consider the following program (which I would appreciate if you add as a test, this showcases the semantics not explored by tests so far, I think):

```
//# publish
module 0x42::m {

fun foo(x: &mut u64): u64 {
    *x = *x + 1;
    10
}

fun main(): u64 {
    let x = 0;
    for (i in 0..foo(&mut x)) {};
    x
}
}

//# run 0x42::m::main
```

We return `11` in this case: this is also a divergence from rust (which only calls `foo` once in an equivalent program), and might be surprising behavior for many (that the function gets evaluated multiple times, so for efficiency, they would need to manually create a temporary result).

So, I think we should discuss all the divergent (from rust) behaviors here to make sure we nail the semantics of a construct that will likely be used quite heavily. It is okay to diverge from other languages, but we need to have good arguments. It would be difficult to go back on some of these choices.

File: third_party/move/move-compiler-v2/tests/checking/typing/while_condition.move
Comment (wrwg) [reply third_party/move/move-compiler-v2/tests/checking/typing/while_condition.move:4]: Actually it does, see the changed .exp file.

File: third_party/move/move-compiler/src/parser/syntax.rs
Comment (wrwg) [inline third_party/move/move-compiler/src/parser/syntax.rs:1230]: There is `consume_identifier` which does the check for the token kind and value, you can use it here and elsewhere.

**Review by wrwg — COMMENTED** [review]
Looks good, nits only

File: third_party/move/move-compiler/transactional-tests/tests/control_flow/nested_for_while_loops.move
Comment (brmataptos) [inline third_party/move/move-compiler/transactional-tests/tests/control_flow/nested_for_while_loops.move:11]: This is an odd test case.  No matter how many times the first loop is traversed, y==1 here.

File: third_party/move/move-compiler/transactional-tests/tests/control_flow/nested_for_while_loops.move
Comment (brmataptos) [inline third_party/move/move-compiler/transactional-tests/tests/control_flow/nested_for_while_loops.move:14]: And a lot of incorrect implementations of this for loop will still leave y==11 at the end.

File: third_party/move/move-compiler/transactional-tests/tests/control_flow/for_type_mismatch.exp
Comment (brmataptos) [inline third_party/move/move-compiler/transactional-tests/tests/control_flow/for_type_mismatch.exp:7]: This set of errors is very confusing to the user.  Can we somehow refine the set of errors we produce?

File: third_party/move/move-compiler/transactional-tests/tests/control_flow/break_continue_for_loop.move
Comment (aalok-t) [reply third_party/move/move-compiler/transactional-tests/tests/control_flow/break_continue_for_loop.move:7]: Added other tests where `i` occurs in loop body.

File: third_party/move/move-compiler/src/parser/syntax.rs
Comment (aalok-t) [reply third_party/move/move-compiler/src/parser/syntax.rs]: Fixed it. Thanks!

File: third_party/move/move-compiler/transactional-tests/tests/control_flow/nested_for_while_loops.move
Comment (aalok-t) [reply third_party/move/move-compiler/transactional-tests/tests/control_flow/nested_for_while_loops.move]: Updated the test.

File: third_party/move/move-compiler/transactional-tests/tests/control_flow/nested_for_while_loops.move
Comment (aalok-t) [reply third_party/move/move-compiler/transactional-tests/tests/control_flow/nested_for_while_loops.move]: Updated the test.

File: third_party/move/move-compiler/transactional-tests/tests/control_flow/for_type_mismatch.exp
Comment (aalok-t) [reply third_party/move/move-compiler/transactional-tests/tests/control_flow/for_type_mismatch.exp:7]: Resolving this issue needs deeper changes. I will add it as an issue after this PR is merged.

**Comment by brmataptos** [comment]
Can we have a few tests for potential existing for usages, e.g.:
```
module 0x42::test {
  fun  for(i : u32, j: u32, k:u32) : u32 {
     i + j + k
  }
  fun for_user() : u32 {
      let (i, j, k) = (3, 4, 5);
      let x = for(i, j, k);
      x
  }
}
```

I see that `third_party/move/move-stdlib/nursery/sources/offer.move` has some more, so maybe building the stdlib as part of the normal tests will be useful, too.

**Review by wrwg — APPROVED** [review]

File: third_party/move/move-compiler/src/parser/syntax.rs
Comment (brmataptos) [reply third_party/move/move-compiler/src/parser/syntax.rs:955]: At least add a comment to say why this code isn't the same as the other control flow code.  perhaps:
```
for loop isn't useful in an expression today, so we ignore second result from `parse_for_loop` and never call `parse_binop_exp` (as might be done for some other control constructs above).
```

**Review by brmataptos — APPROVED** [review]

**Review by vineethk — COMMENTED** [review]
@aalok-t: Did you end up addressing the two concerns here? I did not see corresponding test cases I provided added here, or a further discussion.

1. https://github.com/aptos-labs/aptos-core/pull/10711#issuecomment-1836788142
2. https://github.com/aptos-labs/aptos-core/pull/10711#issuecomment-1836803685

I just wanted to make sure we nail the semantics so that we are not stuck with supporting old behavior.

File: third_party/move/move-compiler/transactional-tests/tests/control_flow/break_continue_for_loop.exp
Comment (brmataptos) [inline third_party/move/move-compiler/transactional-tests/tests/control_flow/break_continue_for_loop.exp:6]: This is not a useful test.

**Pull request merged**
